### ISSUE TITLE
Could incorrectly create uninitialized POCO objects within collections during model binding

### ISSUE THREAD

**Issue by dougbu** [aspnet/Mvc#2203]
current `MutableObjectModelBinder` might incorrectly create uninitialized POCO objects when the parent scope (next model type up) gets `ModelMetadata` from somewhere other than the type hierarchy. problem is an incorrect assumption that `ModelMetadata.ContainerType==null` indicates a top-level model.

problem is not urgent because problematic scenarios seem hard to come by. however we should fix this at some point to ease confusion when debugging and avoid problems as the model binding system changes.

TL;DR
see discussion at https://github.com/aspnet/Mvc/pull/2188#discussion_r26705047.  original code comment to reviewers was:
``` c#
// The isTopLevelObject value is wrong whenever a binder (e.g. CollectionModelBinder) uses
// GetMetadataForType(). ...
```

**Comment by danroth27** [comment]
@dougbu Could you please provide a code sample the reproduces the issue so we don't lose track of the impact? Thanks.

**Comment by dougbu** [comment]
@danroth27 probably not.  the reason I mentioned the issue is not urgent is my inability to think of a scenario where problems would be visible.  so it's a _potential_ problem and code smell but not a customer-visible bug.  of course I many be missing a corner case...

**Comment by harshgMSFT** [comment]
See https://github.com/aspnet/Mvc/blob/533474d07cf2e157a34b8e49194d5e5e6274e7b0/test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs#L1982

for a test which hits into this scenario.

**Comment by dougbu** [comment]
@harshgMSFT not at all.  that test's current behaviour is that a collection is created with a `null` element.  no uninitialized POCO instance is created.

the only unusual thing in that test is the (POCO) element type doesn't have a conversion from `string`.

TL;DR
the test behaviour looks odd but is correct: we treat a submission containing "collectionProperty=some value" as if "some value" were a potentially-comma-separated `string` of values.  this also matches what a `<select/>` element with `multiple="multiple"` would submit (in this case, when a single value is selected).

**Comment by harshgMSFT** [comment]
ah I see.

**Comment by Eilon** [comment]
This work will be tracked by whatever unit test results we come up with.
